chore: RFC for Data Volume Insights (#17322)
vectordotdev/vector: rfcs/2023-05-03-data-volume-metrics.md

# RFC 2023-05-02 - Data Volume Insights metrics

Vector needs to be able to emit accurate metrics that can be usefully queried
to give users insights into the volume of data moving through the system.

## Scope

### In scope

- All event metrics within Vector need to emit the estimated JSON size of the
  event. With a consistent method for determining the size it will be easier

>   to accurately compare data in vs data out.

Reviewer: Not _all_, just the volume ones:

* `component_received_event_bytes_total`
* `component_sent_event_bytes_total`
* `component_received_event_total`
* `component_sent_event_total`

- The metrics sent by each sink needs to be tagged with the source id of the
  event so the route an event takes through Vector can be queried.
- Each event needs to be labelled with a `service`. This is a new concept
  within Vector and represents the application that generated the log or

>   metric.

Reviewer: Or trace.

- The service tag and source tag in the metrics needs to be opt in so customers
  that don't need the increased cardinality are unaffected.

### Out of scope

- A separate metric, `component_sent_bytes_total` that indicates network bytes

>   sent by Vector is not considered here.

Reviewer: And `component_received_bytes_total`.

## Pain

Currently it is difficult to accurately gauge the volume of data that is moving
through Vector. It is difficult to query where data being sent out has come
from.

> ## Proposal

Reviewer: Could we list out the new timeseries/tags that we be published? I think that'll help reviewers understand this change.

Author: I'm not sure I understand what you mean? Do you mean a complete list of all the metric names together with their associated  tags that Vector publishes?

Author: > Could we list out the new timeseries/tags that we be published? I think that'll help reviewers understand this change.

I'm not sure I understand what you mean? Do you mean a complete list of all the metric names together with their associated tags that Vector publishes?

Reviewer: Ah, yeah, basically that. I think that would help readers understand exactly what is changing. For example something like:

```
component_received_event_bytes_total{component_id, component_name, component_type}
```

changing to

```
component_received_event_bytes_total{component_id, component_name, component_type, service, source_id}
```

For each of the metrics that will be updated here.

### User Experience

Global config options will be provided allowing the `service` tag and the
`source` tag to be specified. For example:

```yaml

> metrics:

Reviewer: ```suggestion
internal_metrics:
```

I think calling it `metrics` could get confusing when trying to differentiate from metrics that Vector is processing vs. internal metrics. An alternative name could be `telemetry`.

Author: I think calling it `internal_metrics` might confuse things with the `internal_metrics` source. I'll change it to `telemetry`.

  tags:
    service: service

>     source: input

Reviewer: I assume this is a map of where to find the field? Should we use `log_schema` instead for the mapping? Also, how does this overlap with "semantic meaning" and the associated VRL functions?

Author: No. This is the name of the tag that will be attached to the metric emitted by Vector. I'll make this more clear in the RFC.

```

The default will be to not emit these tags.

### Implementation

#### Metric tags

*service* - to attach the service, we need to add a new meaning to Vector - *service*. Any sources that
            receive data that could potentially be considered a service will need to indicate which field
            means `service`.
            This work has largely already been done with the LogNamespacing work, so it will be trivial to add

>             this new field.

Reviewer: I assume this will use the same log schema concept we added?

The Amqp sink currently emits this value as the length of the binary data that is sent. By the time the data has
reached the code where the `component_sent_event_bytes_total` event is emitted, that event has been encoded
and the actual estimated JSON size has been lost. The sink will need to be updated so that when the event is
encoded, the encoded event together with the pre-encoded JSON bytesize will be sent to the service where the
event is emitted.

The Kafka sink also currently sends the binary size, but it looks like the estimated JSON bytesize is easily
accessible at the point of emitting, so would not need too much of a change.

To ensure that the correct metric is sent in a type-safe manner, we will wrap the estimated JSON size in a
newtype:

```rust
pub struct JsonSize(usize);
```

The `EventsSent` metric will only accept this type.

## Rationale

The ability to visualize data flowing through Vector will allow users to ascertain
the effectiveness of the current use of Vector. This will enable users to
optimise their configurations to make the best use of Vector's features.

## Drawbacks

The additional tags being added to the metrics will increase the cardinality of

> those metrics if they are enabled.

Reviewer: Also curious if this breaks some of the event registration work we did with the need to add this new dynamic tag. cc/ @bruceg (who I also tagged as reviewer)

Author: It looks like it will.

Reviewer 2: It absolutely will, ouch. Given that registered events have become a correctness issue due to expiring metrics, I think we will have to resolve that breakage. I see only one way out, and that is to introduce a registered event caching layer that is keyed on the source/service identifiers. Obviously, that adds more complexity and interacts negatively with expiring metrics.

## Prior Art

## Alternatives

We could use an alternative metric instead of estimated JSON size.

- *Network bytes* This provides a more accurate picture of the actual data being received
  and sent by Vector, but will regularly produce different sizes for an incoming event
  to an outgoing event.
- *In memory size* The size of the event as held in memory. This may be more accurate in
  determining the amount of memory Vector will be utilizing at any time, will often be
  less accurate compared to the data being sent and received which is often JSON.

## Outstanding Questions

- Should the default be to not emit tags? Could the default be different for OP users and standard Vector
  users?
- What should the source be tagged as for events that go through a transform that could potentially
  combine one event from multiple sources - eg. `reduce`?
- What should the source be for events that come from no source (emitted by the `lua` transform)?
- Not all sources will have a `service`. Should we default the service to a value, emit a null service tag

>   or not include the tag at all if there is no service?

Reviewer 3: > Should the default be to not emit tags (source and service)? Could the default be different for OP users and standard Vector users?

This has product implications so Barry can speak to those aspecta of it, but seems we can enforce it on the OP side to make sure it’s consistent for customers (and can be relied upon in the API/UI).  This would be a settings that we can switch on/off on internal metrics?  So when we setup this pipeline internal pipeline we can insure the tags are being set?
https://github.com/DataDog/observability-pipelines-worker/blob/main/src/configuration/internal_pipeline.rs#L44-L52
I would say having a setting for enabling each source and service would give us flexibility.
> What should the source be tagged as for events that go through a transform that could potentially combine one event from multiple sources - eg. reduce?

Let Barry give the product take on this.  Right.  Are there other scenarios you can think of?  For reduce I think I would take the first source id that enters the transaction or some other stable logic (based on what’s easy to implement).  It seems we could recommend not crossing source boundaries if people want better visibility for these cases.
> What should the source be for events that come from no source (emitted by the lua transform)?

Are there other examples of this?  Lua is probably not worth spending a lot of time on on.  Could we just use an internal tag value like   _no_source  or _unknown ? At least we could bucket them and handle them in some way in the API/UI.
> Not all sources will have a service. Should we default the service to a value, emit a null service tag  or not include the tag at all if there is no service?

Similar to above, my :twocents: is that we should use an internal tag value like _no_service (or _none) instead of null tags.  But maybe there’s a convention or best practice around this.  Null feels like some code forgot to set the value or something rather it being a deliberate choice.

Reviewer 3: from @barieom 

> Should the default be to not emit tags (source and service)? Could the default be different for OP users and standard Vector users?
This has product implications so Barry can speak to those aspect of it, but seems we can enforce it on the OP side to make sure it’s consistent for customers (and can be relied upon in the API/UI).  This would be a settings that we can switch on/off on internal metrics?  So when we setup this pipeline internal pipeline we can insure the tags are being set?
https://github.com/DataDog/observability-pipelines-worker/blob/main/src/configuration/internal_pipeline.rs#L44-L52

Yeah having it turned on by default for OP users and turned off by default for standard Vector users, as it may have some performance implications make sense. Speaking of, could you share any numbers around performance regressions when this gets implemented?
Also to add more context on how this feature would evolve - users should be able to define whatever key_field they want to bucket their volume insights by in OP (likely in the bootstrap.yaml or elsewhere), rather than supporting this just for the .service field like we’re doing now/in V1
> What should the source be tagged as for events that go through a transform that could potentially combine one event from multiple sources - eg. reduce?

Like Steve is asking, are scenarios outside of reduce where this could happen?
In the case of reduce, this seems like an edge case and not something many users would do (cc/ 
[@Ari Adair](https://dd.slack.com/team/U02MM3ABX2L) [@Jesse Szwedko](https://dd.slack.com/team/U01NEAWEB4G)
 in case you’ve seen otherwise and if you can share examples if so), but defaulting to source ID with the most number of events reduced into the single event then defaulting to first source id makes sense
> What should the source be for events that come from no source (emitted by the lua transform)?
Are there other examples of this?  Lua is probably not worth spending a lot of time on on.  Could we just use an internal tag value like   _no_source  or _unknown ? At least we could bucket them and handle them in some way in the API/UI.

+1 to Steve here. _unknown or from_<transform_ID> make sense
> Not all sources will have a service. Should we default the service to a value, emit a null service tag  or not include the tag at all if there is no service?
Similar to above, my :twocents: is that we should use an internal tag value like _no_service (or _none) instead of null tags.  But maybe there’s a convention or best practice around this.  Null feels like some code forgot to set the value or something rather it being a deliberate choice.

Also +1 to Steve

## Plan Of Attack

Incremental steps to execute this change. These will be converted to issues after the RFC is approved:

- [ ] Submit a PR with spike-level code _roughly_ demonstrating the change.
- [ ] Incremental change #1
- [ ] Incremental change #2
- [ ] ...

> Note: This can be filled out during the review process.

Reviewer: Just a note to fill this out.

  compare data in vs data out.
  - `component_received_event_bytes_total`
  - `component_sent_event_bytes_total`
  - `component_received_event_total`
  - `component_sent_event_total`
- The metrics sent by each sink needs to be tagged with the source id of the
  event so the route an event takes through Vector can be queried.
- Each event needs to be labelled with a `service`. This is a new concept
  within Vector and represents the application that generated the log,
  metric or trace.
- The service tag and source tag in the metrics needs to be opt in so customers
  that don't need the increased cardinality are unaffected.

### Out of scope

- Separate metrics, `component_sent_bytes_total`  and `component_received_bytes_total`
  that indicate network bytes sent by Vector are not considered here.

## Pain

Currently it is difficult to accurately gauge the volume of data that is moving
through Vector. It is difficult to query where data being sent out has come
from.

## Proposal

### User Experience

Global config options will be provided allowing the name of the `service` tag and the
`source` tag to be specified. For example:

```yaml

> telemetry:

Reviewer: Calling this `telemetry` actually gives me the idea that this should apply to emitted logs too. Curious what you think of that? As a user, I might expect it.

Reviewer 3: Yeah, tagging the logs is a good thought 👍  I don't know what the effort level is but I could see users wanting that.

Author: I've added this to the future improvements section.

  tags:

>     service: theservice

Reviewer: This seems slightly awkward for the, presumably most common, case of `service: service` but it's not terrible.

Author: Would it make more sense to just have this field a boolean and not allow the actual tag name to be changed?

Reviewer: Ah, yeah, a boolean makes sense.

Reviewer 3: Agree on the boolean 👍 Until we think about doing fancier dynamic tags, might be best to keep this simple.

    source: theinput
```

This will cause Vector to emit a metric like (note the last two tags):

```statds
vector.component_sent_event_bytes_total:123|c|#component_id:out,component_kind:sink,component_name:out,component_type:console,host:machine,theservice:somekindofservice,theinput:stdin
```

The default will be to not emit these tags.

### Implementation

#### Metric tags

> **service** - to attach the service, we need to add a new meaning to Vector - *service*. Any sources that

Reviewer: I'm also wondering how this intersects with semantic meanings. It seems like "service" is a new semantic meaning so that users could set this in VRL if they wanted? Is that part of the plan?

> **service** - to attach the service, we need to add a new meaning to Vector - *service*. Any sources that

Author: I've added a bit here to indicate VRL can be used for this in time.

              receive data that could potentially be considered a service will need to indicate which field
              means `service`.
              This work has largely already been done with the LogNamespacing work, so it will be trivial to add
              this new field.

> **source** - A new field will be added to the [Event metadata][event_metadata] - `Arc<ComponentId>` that will indicate the source

Author: There's nothing stopping `OutputId` from being serializable, so it should be fine to be stored in the buffer. 

If the event is loaded from the buffer and it is pointing to a source that no longer exists, given that it is just an identifier it won't cause any issues.

             of the event.

We will need to do an audit of all components to ensure the bytes emitted for the `component_received_event_bytes_total`
and `component_sent_event_bytes_total` metrics are the estimated JSON size of the event.

These tags will be given the name that was configured in [User Experience](#user-experience).

Transforms `reduce` and `aggregate` combine multiple events together. In this case the `source` and `service`
of the first event will be taken.

> If there is no `source` specified (the event was created by the `lua` trnasform) - a source of `_no_source` will

Reviewer: This is a nit since it is easy to adjust later, but I might opt for `""` if the tag isn't there. Or `-`. `_no_service` is a bit verbose when the tag is already called `service`.

Author: I'll go with `-`.

As an example, from a cursory glance over a couple of sinks:

The Amqp sink currently emits this value as the length of the binary data that is sent. By the time the data has
reached the code where the `component_sent_event_bytes_total` event is emitted, that event has been encoded
and the actual estimated JSON size has been lost. The sink will need to be updated so that when the event is
encoded, the encoded event together with the pre-encoded JSON bytesize will be sent to the service where the
event is emitted.

The Kafka sink also currently sends the binary size, but it looks like the estimated JSON bytesize is easily
accessible at the point of emitting, so would not need too much of a change.

To ensure that the correct metric is sent in a type-safe manner, we will wrap the estimated JSON size in a
newtype:

```rust
pub struct JsonSize(usize);
```

The `EventsSent` metric will only accept this type.

## Rationale

The ability to visualize data flowing through Vector will allow users to ascertain
the effectiveness of the current use of Vector. This will enable users to
optimise their configurations to make the best use of Vector's features.

> ## Drawbacks

Reviewer: These metrics should be measuring "in-memory bytes" at the moment so I think the switch to estimated JSON size should only be an improvement. It'd surprise me if users were using in-memory bytes for anything. It is a breaking change that we should call out though (and potentially do the normal deprecate, default, remove, cycle). We could use the new `telemetry` top-level field to do that by adding a "flag" to opt into the new measure. If it is hard, though, we can evaluate whether it is necessary to do the normal cutover for this feature.

> ## Drawbacks

Reviewer 4: Okay, that makes more sense then. Seems like the ones I was thinking of (kafka) are being considered bugs and should be changed anyway. I agree, JSON size is probably better than in-memory. So I don't think this needs to be called out here.

> ## Drawbacks

Author: Also `component_sent_bytes_total`  and `component_received_bytes_total` which measure network bytes are not being changed.

The additional tags being added to the metrics will increase the cardinality of
those metrics if they are enabled.

We will lose the ability to preregister the metrics since the tags will need to be
dynamic. This will cause a noticable, but likely negligible performance loss.

## Prior Art

## Alternatives

We could use an alternative metric instead of estimated JSON size.

- *Network bytes* This provides a more accurate picture of the actual data being received
  and sent by Vector, but will regularly produce different sizes for an incoming event
  to an outgoing event.
- *In memory size* The size of the event as held in memory. This may be more accurate in
  determining the amount of memory Vector will be utilizing at any time, will often be
  less accurate compared to the data being sent and received which is often JSON.

## Outstanding Questions

## Plan Of Attack

Incremental steps to execute this change. These will be converted to issues after the RFC is approved:

- [ ] Add the `source` field to the Event metadata to indicate the source the event has come from.

> - [ ] Update the Volume event metrics to take a `JsonSize` value. Use the compiler to ensure all metrics

Reviewer: Does this step imply the audit of components?

Author: Yeah. Once the metrics take `JsonSize`, Vector will not compile until all the appropriate places have been updated.

Reviewer 2: We could potentially also change the `EstimatedJsonEncodedSizeOf` trait to return a `JsonSize` to make that more obvious as well.

  event. With a consistent method for determining the size it will be easier to accurately
  compare data in vs data out.
  - `component_received_event_bytes_total`
  - `component_sent_event_bytes_total`
  - `component_received_event_total`
  - `component_sent_event_total`
- The metrics sent by each sink needs to be tagged with the source id of the
  event so the route an event takes through Vector can be queried.
- Each event needs to be labelled with a `service`. This is a new concept
  within Vector and represents the application that generated the log,
  metric or trace.
- The service tag and source tag in the metrics needs to be opt in so customers
  that don't need the increased cardinality are unaffected.

### Out of scope

- Separate metrics, `component_sent_bytes_total`  and `component_received_bytes_total`
  that indicate network bytes sent by Vector are not considered here.

## Pain

Currently it is difficult to accurately gauge the volume of data that is moving
through Vector. It is difficult to query where data being sent out has come
from.

## Proposal

### User Experience

Global config options will be provided allowing the name of the `service` tag and the

> `source` tag to be specified. For example:

Reviewer 3: I think we need to update this as well.  I think the global options are just toggling them on/off.  The names are static for now.

```yaml
telemetry:
  tags:

>     service: theservice
>     source: theinput

Reviewer: ```suggestion
    service: true
    source_id: true
```

Suggesting based on https://github.com/vectordotdev/vector/pull/17322/files#r1190277614

```

This will cause Vector to emit a metric like (note the last two tags):

```statds
vector.component_sent_event_bytes_total:123|c|#component_id:out,component_kind:sink,component_name:out,component_type:console,host:machine,theservice:somekindofservice,theinput:stdin
```

The default will be to not emit these tags.

### Implementation

#### Metric tags

> **service** - to attach the service, we need to add a new meaning to Vector - 

Reviewer: Could we include an example showing that users can modify this meaning (e.g. set it to point to a different field than `service`)?

Reviewer: I believe they already exist? The `set_semantic_meaning` function.

Author: Ah yes. Just undocumented.

              `service`. Any sources that receive data that could potentially 
              be considered a service will need to indicate which field means 
              `service`. This work has largely already been done with the 
              LogNamespacing work, so it will be trivial to add this new field.

**source** - A new field will be added to the [Event metadata][event_metadata] - 
             `Arc<OutputId>` that will indicate the source of the event. 
             `OutputId` will need to be serializable so it can be stored in the

>              disk buffer. If the event is loaded from the buffer and it is pointing
>              to a source that no longer exists, given that it is just an identifier
>              it won't cause any issues.

Reviewer 2: This isn't specific to buffers either. The source could be unloaded while an event is still in flight.
```suggestion
             disk buffer. Since this field is just an identifier, it can still be used
             even if the source no longer exists when the event is consumed
             by a sink.
```

We will need to do an audit of all components to ensure the
bytes emitted for the `component_received_event_bytes_total` and
`component_sent_event_bytes_total` metrics are the estimated JSON size of the

> event.

Reviewer: I think it is still (incorrectly) used for building batches (#10020). It could be useful for https://github.com/vectordotdev/vector/issues/8679 but we could also wait until we implement that to re-add it if needed.

These tags will be given the name that was configured in [User Experience]
(#user-experience).

Transforms `reduce` and `aggregate` combine multiple events together. In this
case the `source` and `service` of the first event will be taken.

If there is no `source` specified (the event was created by the `lua` transform)
- a source of `-` will be emitted.

> If there is no `service` available, a service of `-` will be emitted.

Reviewer 2: I'd like to see a rationale for mandating these stub tag values as opposed to just eliding the tags when not present.

Reviewer: Agreed, we should call that out explicitly here, but we wouldn't want to publish without the tag since that would indicate that the metric represents the aggregate value across all `service`s.

Despite being in the [Component Spec][component_spec], not all sinks currently
conform to this.

As an example, from a cursory glance over a couple of sinks:

The Amqp sink currently emits this value as the length of the binary
data that is sent. By the time the data has reached the code where the
`component_sent_event_bytes_total` event is emitted, that event has been
encoded and the actual estimated JSON size has been lost. The sink will need
to be updated so that when the event is encoded, the encoded event together
with the pre-encoded JSON bytesize will be sent to the service where the event
is emitted.

The Kafka sink also currently sends the binary size, but it looks like the
estimated JSON bytesize is easily accessible at the point of emitting, so would
not need too much of a change.

To ensure that the correct metric is sent in a type-safe manner, we will wrap
the estimated JSON size in a newtype:

```rust
pub struct JsonSize(usize);
```

The `EventsSent` metric will only accept this type.

### Registered metrics

It is currently not possible to have dynamic tags with preregistered metrics.

We will need to introduce a registered event caching layer that will register

> and cache new events keyed on the tags that are sent to it.

Author: Yes we do need to do this, otherwise we risk the event counts expiring and resetting to zero.

  to an outgoing event.
- *In memory size* The size of the event as held in memory. This may be more accurate in
  determining the amount of memory Vector will be utilizing at any time, will often be
  less accurate compared to the data being sent and received which is often JSON.

## Outstanding Questions

## Plan Of Attack

Incremental steps to execute this change. These will be converted to issues after the RFC is approved:

- [ ] Add the `source` field to the Event metadata to indicate the source the event has come from.
- [ ] Update the Volume event metrics to take a `JsonSize` value. Use the compiler to ensure all metrics
      emitted use this. The `EstimatedJsonEncodedSizeOf` trait will be updated return a `JsonSize`.
- [ ] Add the Service meaning. Update any sources that potentially create a service to point the meaning
      to the relevant field.
- [ ] Introduce an event caching layer that caches registered events based on the tags sent to it.
- [ ] Update the emitted events to accept the new tags - taking the `telemetry` configuration options
      into account.

## Future Improvements

> - Logs emitted by Vector should also be tagged with `source_id` and `service`.

Reviewer: Do we know how much effort this would be? If it is on the order of a few days it might be nice to do as part of this work since I can see users expecting it. If it'll take weeks then we can defer it.

Author: It's more than we can fit in to the deadline, we agreed to defer for now.

those metrics if they are enabled.

## Prior Art

## Alternatives

We could use an alternative metric instead of estimated JSON size.

- *Network bytes* This provides a more accurate picture of the actual data being received
  and sent by Vector, but will regularly produce different sizes for an incoming event
  to an outgoing event.
- *In memory size* The size of the event as held in memory. This may be more accurate in
  determining the amount of memory Vector will be utilizing at any time, will often be
  less accurate compared to the data being sent and received which is often JSON.

## Outstanding Questions

## Plan Of Attack

Incremental steps to execute this change. These will be converted to issues after the RFC is approved:

- [ ] Add the `source` field to the Event metadata to indicate the source the event has come from.
- [ ] Update the Volume event metrics to take a `JsonSize` value. Use the compiler to ensure all metrics
      emitted use this. The `EstimatedJsonEncodedSizeOf` trait will be updated return a `JsonSize`.
- [ ] Add the Service meaning. Update any sources that potentially create a service to point the meaning
      to the relevant field.

> - [ ] Introduce an event caching layer that caches registered events based on the tags sent to it.

Reviewer: I'd like to add benchmarking as an explicit part of the plan of action here. I think it'll be important to understand the performance impact.

Author: Have added.
